[stdlib] String from code units API

zwaldowski opened the pull request:
apple/swift-evolution#101 seems to have died on the vine. Since the addition was non-controversial, and the code is already complete, I figured I'd make some noise so we can go ahead and get the evolution proposal started/completed; in the meantime, I can get some eyes on the code changes. Cheers!

glessard commented:
I think both the StringDecodeTests and CStringTests should move over from `test/1_stdlib/NSStringAPI.swift` to `validation-test/stdlib/String.swift`, for testing on Linux.  (The `NSStringAPI` file requires objc interop).

moiseev commented:
The proposal hasn't died. It's just that at the moment there are larger ones being discussed. Your proposal will get reviewed eventually, I'm keeping an eye on it. Sorry it takes so long.

zwaldowski commented:
Good point. Added to my to-do list. I don't have a way to run the tests
on Linux readily, but I'll start pulling that together.

Zachary Waldowski zach@waldowski.me


On Wed, Jan 27, 2016, at 02:30 PM, Guillaume Lessard wrote:
> I think both the StringDecodeTests and CStringTests should move over
> from test/1_stdlib/NSStringAPI.swift to validation-
> test/stdlib/String.swift, for testing on Linux.  (The NSStringAPI file
> requires objc interop).


> — Reply to this email directly or view it on GitHub[1].


>



Links:

  1. https://github.com/apple/swift/pull/1109#issuecomment-175811140

zwaldowski commented:
My apologies if I seemed pushy, I just didn't want it to turn into an
"only the squeaky wheel gets the grease" thing. I understand there's a
lot more important stuff going on.

On Wed, Jan 27, 2016, at 02:34 PM, Maxim Moiseev wrote:
> The proposal hasn't died. It's just that at the moment there are
> larger ones being discussed. Your proposal will get reviewed
> eventually, I'm keeping an eye on it. Sorry it takes so long.


> — Reply to this email directly or view it on GitHub[1].


>



Links:

  1. https://github.com/apple/swift/pull/1109#issuecomment-175812729

moiseev commented:
@zwaldowski We have https://ci.swift.org now, it will happily run all your tests on multiple platforms. ;-)

glessard commented:
@zwaldowski https://github.com/apple/swift/pull/1131 caused a conflict with this. I moved the tests of non-Foundation API that were in `test/1_stdlib/NSStringAPI.swift` to `test/1_stdlib/StringAPI.swift`.

zwaldowski commented:
@glessard Thanks for the heads up! I've resolved the merge but will wait to push until master is fixed and I can verify the changes. Cheers!

zwaldowski committed:
Expose String-from-code-units as public initializers

String._fromCodeUnitSequence(_:input:) is replaced with
String.init?(codeUnits:encoding:).

String._fromCodeUnitSequenceWithRepair(_:input:) is replaced with
String.fromCodeUnitsWithRepair(_:encoding:).

zwaldowski committed:
Revise with fuller range of code units constructors

zwaldowski committed:
Introduce new-style CString constructors

zwaldowski committed:
Adopt String(validatingCodeUnits:as:) where optional is already returned

zwaldowski committed:
Replace uses of deprecated CString constructors

zwaldowski committed:
Elide other uses of String.Type._fromWellFormedCodeUnitSequence(_:input:)

zwaldowski commented:
Rebased against master, stdlib tests past locally. Will run the benchmarks for it shortly. Do I have to re-open the PR to kick CI into gear?

Something of note, at least as far as 2.2 relevance — the new non-nil pointer requirement for the C-string constructors was painful to diagnose on test runs, and I'm sure there are plenty more lurking throughout the stdlib (one hotspot was Reflection). I don't know what the policy is for invariants changing between versions — would it be reasonable to continue to eat null C-strings until we get pointer optionality in 3?

moiseev commented:
@swift-ci Please test

moiseev commented:
It is not just an invariant change of an existing method, it is a completely new API. So new invariants are part of the deal, in my opinion.

moiseev commented:
@zwaldowski thanks for keeping it up to date!

zwaldowski commented:
@moiseev: I don't disagree at all, but without modeling the null requirement in the type system, I fear people will blindly migrate in a way that will cause lurking crashers in their code.

zwaldowski commented:
Ah… seems corelibs Foundation is using the prior `_fromCodeUnitSequence` version. Might need to restore that with a deprecation attribute for their benefit.

zwaldowski commented on stdlib/public/core/CString.swift:
```diff
@@ -15,19 +15,50 @@
   /// Returns `nil` if the `CString` is `NULL` or if it contains ill-formed
   /// UTF-8 code unit sequences.
   @warn_unused_result
+  @available(*, deprecated, message="Use String(validatingCString:) initializer")
```
Is this the right recommendation to make? If someone chose `fromCString(_:)` just for brevity, would it make more sense to recommend `init(cString:)` instead?

moiseev replied:
I don't know if there is any limit to the length of the message, so... maybe something along the lines of "Use String(cString:) or String(validatingCString:) with nullness check instead".
One unrelated note: in our design, argument for a failing initializer is called `validatingUtf8`, as `validatingCString` does not really reflect what gets validated (which is the UTF8 code units). What do you think?

zwaldowski replied:
Ah, fair point, re: recommending both.

I feel `cString: UnsafePointer<CChar>` unambiguously describes what people are used to, whereas reusing just `utf8` doesn't indicate the difference in semantics from the `codeUnits` versions (i.e., `strlen` and null avoidance). Plus it gets into initialism concerns, which is a dog I have no fight in. :stuck_out_tongue_winking_eye:

zwaldowski replied:
I feel like further conversations of the name (since it's a good route to go down) should be put on the Evolution thread.

zwaldowski committed:
Recommend both alternatives for String.Type.fromCString(_:)

zwaldowski commented:
Closing based on Evolution discussion.
